File: cmd/parser/bezier.go
package main

import (
	"gonum.org/v1/plot"
	"gonum.org/v1/plot/plotter"
)

type Vector2 struct {
	X float64
	Y float64
}

func LinearInterpolation(start, end Vector2, t float64) Vector2 {
	return Vector2{
		X: start.X + (end.X-start.X)*t,
		Y: start.Y + (end.Y-start.Y)*t,
	}
}

func BezierInterpolation(p0, p1, p2 Vector2, t float64) Vector2 {
	intermediateA := LinearInterpolation(p0, p1, t)
	intermediateB := LinearInterpolation(p1, p2, t)
	return LinearInterpolation(intermediateA, intermediateB, t)
}

func DrawBezier(p *plot.Plot, p0, p1, p2 Vector2, resolution int) error {
	points := make(plotter.XYs, resolution+1)
	points[0] = plotter.XY{X: p0.X, Y: p0.Y}

	for i := 1; i <= resolution; i++ {
		t := float64(i+1) / float64(resolution)
		nextPointOnCurve := BezierInterpolation(p0, p1, p2, t)

		points[i].X = nextPointOnCurve.X
		points[i].Y = nextPointOnCurve.Y
	}

	line, err := plotter.NewLine(points)
	if err != nil {
		return err
	}
	p.Add(line)

	return nil
}

File: parser/bezier.py
class Point:
    def __init__(self, x: int, y: int, on_curve: bool):
        self.x = x
        self.y = y
        self.on_curve = on_curve

    def __repr__(self):
        return f"({self.x}, {self.y})"


def DrawLine(p1: Point, p2: Point, ax):
    ax.plot([p1.x, p2.x], [p1.y, p2.y], "k-")


def DrawPoint(p: Point, ax):
    ax.plot(p.x, p.y, "ro", markersize=3)


def LinearInterpolation(start: Point, end: Point, time: float) -> Point:
    return Point(
        x=start.x + (end.x - start.x) * time,
        y=start.y + (end.y - start.y) * time,
        on_curve=False,
    )


def BezierInterpolation(p0: Point, p1: Point, p2: Point, time: float) -> Point:
    intermediateA = LinearInterpolation(p0, p1, time)
    intermediateB = LinearInterpolation(p1, p2, time)
    return LinearInterpolation(intermediateA, intermediateB, time)


def DrawBezier(p0: Point, p1: Point, p2: Point, resolution: int, ax) -> None:
    prev_point_on_curve = p0

    for i in range(resolution):
        t = (i + 1.0) / resolution
        next_point_on_curve = BezierInterpolation(p0, p1, p2, t)
        DrawLine(prev_point_on_curve, next_point_on_curve, ax)
        prev_point_on_curve = next_point_on_curve
